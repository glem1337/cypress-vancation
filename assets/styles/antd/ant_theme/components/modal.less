// Modal
@modal-body-padding: 0;
@modal-header-bg: @component-background;
@modal-header-border-color-split: @border-color-split;
@modal-content-bg: @component-background;
@modal-heading-color: @heading-color;
@modal-footer-bg: transparent;
@modal-footer-border-color-split: @border-color-split;
@modal-footer-padding-vertical: 24px;
@modal-footer-padding-horizontal: 24px;
@modal-mask-bg: fade(black, 50%);
