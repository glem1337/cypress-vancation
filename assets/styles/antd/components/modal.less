.ant-modal-wrap {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ant-modal-content {
  @media (max-width: @screen-md - 1px) {
    border-bottom-right-radius: unset;
    border-bottom-left-radius: unset;
  }
}

.ant-modal-close {
  top: 2px;
}

.ant-modal-close-x {
  color: @gray500;
  transition: color 0.3s;

  &:hover {
    color: @black;
  }

  svg {
    display: none;
  }
}
