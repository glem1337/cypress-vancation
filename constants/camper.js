import { range } from 'ramda';

export const CURRENT_MILEAGE = [
  { value: 'Under 50,000', label: 'Under 50,000' },
  { value: '50,001 - 90,000', label: '50,001 - 90,000' },
  { value: '90,001 - 130,000', label: '90,001 - 130,000' },
  { value: '130,001 - 200,000', label: '130,001 - 200,000' },
  { value: '200,001+', label: '200,001+' },
];

export const INSIDE_HEIGHT = [
  {
    value: 'hight_top',
    label: { id: 'addNewCamper.camperDetails.from.highTop' },
    description: { id: 'addNewCamper.camperDetails.from.insideStandingHeight' },
    img: '/images/listing/High-Top.svg',
  },
  {
    value: 'pop_top',
    label: { id: 'addNewCamper.camperDetails.from.popTop' },
    description: { id: 'addNewCamper.camperDetails.from.lowInsideStandingHeightPoppedUp' },
    img: '/images/listing/Pop-Top.svg',
  },
  {
    value: 'low_top',
    label: { id: 'addNewCamper.camperDetails.from.lowTop' },
    description: { id: 'addNewCamper.camperDetails.from.lowInsideStandingHeight' },
    img: '/images/listing/Low-Top.svg',
  },
  {
    value: 'roof_top',
    label: { id: 'addNewCamper.camperDetails.from.roofTopTent' },
    description: { id: 'addNewCamper.camperDetails.from.typicallyLowTopVehicle' },
    img: '/images/listing/Roof-Top-Tent.svg',
  },
];

export const DRIVETRAIN = [
  { value: 'rwd', label: { id: 'shared.RWD' } },
  { value: 'fwd', label: { id: 'shared.FWD' } },
  { value: '4x4', label: { id: 'shared.4X4' } },
  { value: 'wd', label: { id: 'shared.WD' } },
];

export const TRANSMISSION = [
  { value: 'automatic', label: { id: 'shared.automatic' } },
  { value: 'manual', label: { id: 'shared.manual' } },
];

export const FUEL_TYPE = [
  { value: 'diesel', label: { id: 'shared.diesel' } },
  { value: 'gas', label: { id: 'shared.gas' } },
  { value: 'electric', label: { id: 'shared.electric' } },
];

export const START_RANGE_AGE_CAMPER = 1970;

export const START_RANGE_LENGTH_CAMPER = 10;
export const START_RANGE_LENGTH_CAMPER_FLOAT = 16;

export const FINISH_RANGE_LENGTH_CAMPER = 40;
export const FINISH_RANGE_LENGTH_CAMPER_FLOAT = 30;

export const MAX_SEAT_SLEEP_PLACE = 20;

export const MIN_SEAT_SLEEP_PLACE = 1;

export const FORM_VALIDATION = {
  MAX_LISTING_NAME: 74,
  MAX_LISTING_DESCRIPTION: 2000,
  MAX_PHOTOS_COUNT: 30,
  MAX_PHOTOS_SIZE: 10485760,
  IMAGE_TYPES: 'image/jpeg, image/png, image/jpg',
  MAX_CUSTOM_FEES_COUNT: 5,
};

export const TRIP_FEES = {
  MILEAGE_VALUES: {
    UNLIMITED: 'unlimited',
    LIMITED: 'limited',
  },
  GENERATOR_VALUES: {
    UNLIMITED: 'unlimited',
    LIMITED: 'limited',
  },
  FREQUENCY_OPTIONS: {
    PER_DAY: {
      LABEL: { id: 'addNewCamper.tripFees.perDay' },
      VALUE: 'per_day',
    },
    PER_HOUR: {
      LABEL: { id: 'addNewCamper.tripFees.perHour' },
      VALUE: 'per_hour',
    },
    PER_EACH: {
      LABEL: { id: 'addNewCamper.tripFees.perEach' },
      VALUE: 'per_each',
    },
  },
  DUMPING_VALUES: {
    DUMPING: 0,
    FLUE: 20,
    LATE_DROP_OFF: 20,
  },
};

export const POLICIES_FORM_VALUES = {
  INSTANT_BOOK: 'instant_book',
  REVIEW: 'review',
  EASY_GOING: 'easy_going',
  FIRM_BUT_FAIR: 'firm_but_fair',
  BY_THE_BOOK: 'by_the_book',
};

export const POLICIES_FORM_DEFAULT_VALUES = {
  BOOKING_APPROVAL_POLICY: POLICIES_FORM_VALUES.INSTANT_BOOK,
  CANCELLATION_POLICY: POLICIES_FORM_VALUES.EASY_GOING,
  REQUEST_NOTICE: false,
  AUTO_BLOCKED_DAYS: 1,
  DEPOSIT: 2500,
};

export const CANCELLATION_POLICY_DESCRIPTION_BY_VALUE = {
  [POLICIES_FORM_VALUES.EASY_GOING]: {
    title: {
      id: 'addNewCamper.policies.easyTitle',
    },
    description: {
      id: 'addNewCamper.policies.easyDescription',
    },
  },
  [POLICIES_FORM_VALUES.FIRM_BUT_FAIR]: {
    title: {
      id: 'addNewCamper.policies.firmTitle',
    },
    description: {
      id: 'addNewCamper.policies.firmDescription',
    },
  },
  [POLICIES_FORM_VALUES.BY_THE_BOOK]: {
    title: {
      id: 'addNewCamper.policies.bookTitle',
    },
    description: {
      id: 'addNewCamper.policies.bookDescription',
    },
  },
};

export const POLICIES_FORM_NOTICE_OPTIONS = range(1, 11)
  .map((number) => ({
    value: number,
    label: {
      id: 'addNewCamper.pricing.countOfNights',
      values: { count: number },
    },
  }));

export const CAMPER_INCLUSION = {
  OWNER: 'owner',
  SPECIFICATIONS_DETAILS: 'specification_detail',
  PRICING_INFO: {
    INDEX: 'camper_calendar.pricing_info',
    WEEK_NIGHT_PRICE: 'camper_calendar.pricing_info.week_night_price',
  },
  DELIVERY_INFORMATION: 'delivery_information',
  CAMPER_RULE: 'camper_rule',
  CAMPER_PHOTOS: 'camper_photos',
  INSURANCE_INFO: 'insurance_info',
  TRIP_FEE: {
    INDEX: 'trip_fee',
    TRIP_FEE_MILEAGE: 'trip_fee.trip_fee_mileage',
    TRIP_FEE_GENERATOR: 'trip_fee.trip_fee_generator',
    CUSTOM_FEES: 'trip_fee.custom_fees',
  },
  CAMPER_ADDITION: {
    INDEX: 'camper_addition',
    RESTRICTION_RULE: 'camper_addition.restriction_rule',
    RESTRICTION_ROAD: 'camper_addition.restriction_road',
    TRAVEL_RESTRICTION: 'camper_addition.travel_restriction',
    CUSTOM_RESTRICTION_RULES: 'camper_addition.custom_restriction_rules',
    CUSTOM_TRAVEL_RESTRICTIONS: 'camper_addition.custom_travel_restrictions',
    CUSTOM_RESTRICTION_ROADS: 'camper_addition.custom_restriction_roads',
    CAMPER_TRAVEL_ACCESSORIES: 'camper_addition.camper_travel_accessories',
    CUSTOM_TRAVEL_ACCESSORIES: 'camper_addition.custom_travel_accessories',
    AMENITY_HEALTH_SAFETY_ITEMS: 'camper_addition.amenity_health_safety_items',
    CAMPER_DOCUMENTS: 'camper_addition.camper_documents',
    CAMPER_QUESTIONS: 'camper_addition.camper_questions',
  },
  AMENITIES: {
    INDEX: 'amenities',
    SUB_AMENITIES: 'amenities.sub_amenities',
    CUSTOM_AMENITIES: 'amenities.custom_amenities',
    CONFIGURATION_AMENITY: {
      INDEX: 'amenities.configuration_amenity',
      CONFIGURATION_SUB_AMENITIES: 'amenities.configuration_amenity.configuration_sub_amenities',
      CONFIGURATION_AMENITY_OPTIONS: {
        INDEX: 'amenities.configuration_amenity.configuration_amenity_options',
        CONFIGURATION_SUB_AMENITIES: 'amenities.configuration_amenity.configuration_amenity_options.configuration_sub_amenities',
      },
    },
    AMENITY_OPTIONS: {
      INDEX: 'amenities.amenity_options',
      SUB_AMENITIES: 'amenities.amenity_options.sub_amenities',
    },
  },
  CAMPER_CALENDAR: {
    INDEX: 'camper_calendar',
    PRICING_PERIODS: 'camper_calendar.pricing_periods',
    DISCOUNT_PERIODS: 'camper_calendar.custom_discount_periods',
    BLOCKED_PERIODS: 'camper_calendar.blocked_periods',
    EXTERNAL_CALENDARS: {
      INDEX: 'camper_calendar.external_calendars',
      EVENTS: 'camper_calendar.external_calendars.events',
    },
  },
  VEHICLE: {
    MAKES: 'vehicle_makes',
    MODELS: 'vehicle_makes.vehicle_models',
  },
};

export const CAMPER_PHOTO_DEFAULT = '/images/camper-default.jpg';

export const CAMPER_STATUS = {
  ALL: 'all',
  PUBLISHED: 'published',
  UNPUBLISHED: 'unpublished',
  ON_MODERATION: 'on_moderation',
  DRAFT: 'draft',
  REMOVED: 'removed',
  DEACTIVATED: 'deactivated',
};

export const DEFAULT_MIN_NIGHT_STAY = 2;
