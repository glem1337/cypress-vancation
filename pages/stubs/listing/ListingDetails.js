export { default } from 'views/stubs/listing/ListingDetails';
