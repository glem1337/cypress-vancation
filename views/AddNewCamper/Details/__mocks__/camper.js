export default {
  id: 'a18f7338-c2b3-4cee-be4d-579c0a726c1a',
  name: 'test',
  description: 'desc',
  status: 'draft',
  insurance: 'pending',
  place: null,
  publicId: '111',
  owner: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    type: 'owner',
  },
  specificationDetail: null,
  pricingInfo: null,
  camperPolicy: null,
  deliveryInformation: null,
  camperPhotos: [],
  blockedPeriods: [],
  pricingPeriods: [],
  externalCalendars: [],
};
