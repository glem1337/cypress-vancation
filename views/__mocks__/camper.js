export default {
  id: 'a18f7338-c2b3-4cee-be4d-579c0a726c1a',
  name: 'test',
  description: 'desc',
  status: 'draft',
  insurance: 'pending',
  place: 'Astoria, New York, New York 11102, United States',
  publicId: '111',
  estimatedEarning: 1200.0,
  vehicleTypeIconUrl: 'test.svg',
  vehicleTypeName: 'Modern Van',
  placeId: 'neighborhood.290416',
  longitude: -73.93030000000002,
  latitude: 40.772,
  glamper: true,
  owner: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    account: {
      createdAt: '2021-07-09T13:39:26.587Z',
      drivingLicenceVerified: false,
      email: 'kateryna.sessa+owner@rubygarage.org',
      emailVerified: false,
      id: '12563773-6e8f-4861-8abd-82a78632c6ac',
      user: {
        avatarUrl: null,
        createdAt: '2021-07-09T13:39:26.590Z',
        firstName: 'Kat',
        id: '77015afb-71e3-4b94-a9cb-69a97dc6b358',
        lastName: 'Kat',
        updatedAt: '2021-07-09T13:39:26.590Z',
      },
    },
    type: 'owner',
    avatarUrl: null,
    averageRating: null,
    businessTitle: null,
    campersCount: 25,
    createdAt: 'Member since July 2021',
    description: 'description',
    email: null,
    idVerified: false,
  },
  specificationDetail: {
    drivetrain: '4x4',
    fuelType: 'electric',
    insideHeight: 'low_top',
    length: '67.89',
    modelNaming: 'Clair Weimann',
    name: 'Katelyn Lowe',
    seats: 1,
    sleeps: 1,
    stateRegistred: 'Arkansas',
    transmission: 'automatice',
    year: 1585,
    mileage: 'Under 50,000',
    whoBuiltCamper: 'Vasya',
  },
  pricingInfo: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
  },
  camperRule: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    bookingApprovalPolicy: 'instant_book',
    cancellationPolicy: 'easy_going',
  },
  deliveryInformation: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    pickup: true,
  },
  camperPhotos: [{
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    photoUrl1100: 'photoUrl1100',
  }],
  insuranceInfo: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
  },
  tripFee: {
    id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
    cleaning: 1,
    dumpingFee: 2,
    flue: 3,
    lateDropOff: 4,
    customFees: [
      {
        id: 'id',
        name: 'name',
        price: 5,
        frequency: 'per_day',
      },
    ],
    tripFeeMileage: {
      limit: true,
      available: 6,
      overage: 7,
    },
    tripFeeGenerator: {
      limit: true,
      available: 8,
      overage: 9,
    },
  },
  blockedPeriods: [],
  pricingPeriods: [],
  externalCalendars: [],
  camperCalendar: {
    id: 'e1db81b1-8ad5-4e88-9c99-f5c9020148eb',
    externalCalendars: [],
    blockedPeriods: [
      {
        id: '56ceb535-322f-42c6-8dcd-a88ad59000f3',
        startDate: '2021-08-05',
        endDate: '2021-08-06',
      },
      {
        id: '17a32ec0-a486-4247-910c-6c0c57ba6ec3',
        startDate: '2021-08-25',
        endDate: '2021-08-25',
      },
      {
        id: '2c7bbaef-6b00-494f-b005-6a762e90c785',
        startDate: '2021-07-28',
        endDate: '2021-07-28',
      },
    ],
    pricingPeriods: [
      {
        id: '8aa92a51-ef00-4dc6-8b84-2820869b9278',
        startDate: '2021-07-28',
        endDate: '2021-07-28',
        price: '33.0',
      },
      {
        id: '248ba38e-e18a-467c-996b-36aa4bd73e85',
        startDate: '2021-07-29',
        endDate: '2021-07-30',
        price: '320.0',
      },
    ],
    customDiscountPeriods: [
      {
        id: 'ffb3f0cc-4513-4fdc-81d0-03587232d9ee',
        startDate: '2021-07-27',
        endDate: '2021-07-28',
        weeklyDiscountPercent: 15,
        monthlyDiscountPercent: 23,
      },
      {
        id: 'a070339d-7058-4769-8e16-814084c6bfe3',
        startDate: '2021-07-28',
        endDate: '2021-07-28',
        weeklyDiscountPercent: 15,
        monthlyDiscountPercent: 23,
      },
      {
        id: 'aba16cba-5873-4273-8172-816b276a2db7',
        startDate: '2021-07-28',
        endDate: '2021-07-28',
        weeklyDiscountPercent: 44,
        monthlyDiscountPercent: 44,
      },
      {
        id: '586016b9-d351-4ca1-ac8a-373e7e3b8832',
        startDate: '2021-07-29',
        endDate: '2021-07-29',
        weeklyDiscountPercent: 15,
        monthlyDiscountPercent: 33,
      },
    ],
  },
  camperAddition: {
    amenityHealthSafetyItems: [
      {
        healthSafety: {
          id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
          iconUrl: 'path/to/icon/url',
          name: 'name',
        },
      },
    ],
    restrictionRule: {
      allowPets: false,
      smoking: false,
      festivalApproved: false,
    },
    customRestrictionRules: [
      {
        id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
        active: true,
        name: 'name',
      },
    ],
    travelRestriction: {
      mexico: false,
      canada: true,
      burningMan: false,
    },
    customTravelRestrictions: [
      {
        id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
        active: true,
        name: 'name',
      },
    ],
    restrictionRoad: {
      fourWheelRoad: false,
      offRoad: false,
      snowAndIceRoad: false,
      dirtryRoad: true,
    },
    customRestrictionRoads: [
      {
        id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
        active: true,
        name: 'name',
      },
    ],
    camperQuestions: [
      {
        id: '74a9c587-7b78-4aa5-ba93-ba3d310be127',
        text: 'test',
        required: true,
      },
    ],
    customTravelAccessories: [
      {
        active: true,
        description: '',
        id: 'dd477886-ec22-4da9-ba4c-08627f35fb1a',
        maxAmount: '2.0',
        name: 'Bike Rack',
        price: '1.0',
        priceUnit: 'each',
      },
    ],
  },
};
