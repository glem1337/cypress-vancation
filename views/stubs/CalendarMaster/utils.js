import { findIndex, times } from 'lodash';
import moment from 'moment';

const plainDateFormat = 'YYYY MM DD';

const getCellWidth = (windowWidth) => {
  if (windowWidth < 768) {
    return 50;
  }
  if (windowWidth < 992) {
    return 72;
  }
  return 84;
};

export const getDaysCount = (from, to) => to.diff(from, 'days') + 1;

const toWeekDate = (date) => date.format('ddd DD');

const getBookingLinePosition = (
  from,
  to,
  rangeStart,
  rangeEnd,
  windowWidth,
) => {
  const lineStart = moment.max(rangeStart, from);
  const lineEnd = moment.min(rangeEnd, to);
  const totalBookingHours = moment
    .duration(lineEnd.diff(lineStart))
    .as('hours');
  const percentsPerHour = getCellWidth(windowWidth) / 24;
  const bookingLineWidth = percentsPerHour * totalBookingHours;
  const bookingStartHour = parseInt(moment(lineStart).format('H'), 10);
  const startPositionInCell = percentsPerHour * bookingStartHour;
  return {
    left: startPositionInCell,
    width: bookingLineWidth,
  };
};

export const getHeaderDates = (daysCount, from, today) => {
  const headerDates = [];
  times(daysCount, (idx) => {
    const nextDay = moment(from).add(idx, 'days');
    const formattedNextDay = toWeekDate(nextDay);
    const isToday = moment(today).isSame(nextDay);
    headerDates.push({ formattedNextDay, isToday });
  });
  return headerDates;
};

export const getVanCellsData = (
  rangeStart,
  rangeEnd,
  rangeDaysCount,
  today,
  dates,
  bookings,
  windowWidth,
) => {
  const cells = [];
  times(rangeDaysCount, (idx) => {
    const day = moment(rangeStart).add(idx, 'days');
    const passed = moment(today).diff(day, 'days') > 0;
    cells.push({
      day: day.format(plainDateFormat),
      passed,
    });
  });

  const availableDatesGroups = dates.map(({ from, to, prices }) => {
    const daysCount = getDaysCount(from, to);
    const availableDates = [];
    times(daysCount, (idx) => {
      availableDates.push({
        day: moment(from).add(idx, 'days').format(plainDateFormat),
        price: prices[idx],
      });
    });
    return availableDates;
  });

  availableDatesGroups.forEach((availableDates) => {
    availableDates.forEach((availableDate) => {
      const cellIndex = findIndex(
        cells,
        ({ day }) => day === availableDate.day,
      );
      cells[cellIndex].price = availableDate.price;
    });
  });

  bookings.forEach(({ from, to, details }) => {
    const bookingLinePosition = getBookingLinePosition(
      from,
      to,
      rangeStart,
      rangeEnd,
      windowWidth,
    );
    const beforeRange = moment(from).isBefore(rangeStart);
    const afterRange = moment(to).isAfter(rangeEnd);
    const bookingStartDay = moment
      .max(rangeStart, from)
      .format(plainDateFormat);
    const cellIndexToAddBooking = findIndex(
      cells,
      ({ day }) => day === bookingStartDay,
    );
    cells[cellIndexToAddBooking].booking = {
      position: bookingLinePosition,
      details,
      beforeRange,
      afterRange,
    };
  });

  return cells;
};
