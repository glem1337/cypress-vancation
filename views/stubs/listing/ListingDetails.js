import React from 'react';
import { Form, Col, Input, Row } from 'antd';
import LayoutListing from './LayoutListing';

const { TextArea } = Input;

const ListingDetails = () => (
  <LayoutListing withBackBtn>
    <Form layout="vertical">
      <Row>
        <Col span={24}>
          <h1 className="text-headline mb-8">
            Name and describe your listing
          </h1>
          <p className="mb-16">
            Catchy listing names are key to increasing
            views and bookings. Keep it short and unique.
          </p>
        </Col>
        <Col span={24}>
          <Row>
            <Col lg={16}>
              <Form.Item
                label={<span className="main-input__label">Listing name</span>}
                help={(
                  <p className="main-input__message">
                    <i className="main-input__message-icon icon icon-info" />
                    45/74 character maximum
                  </p>
                )}
              >
                <Input type="text" id="Listing" />
              </Form.Item>
            </Col>
            <Col lg={16}>
              <Form.Item
                label={<span className="main-input__label">Description (optional)</span>}
                help={(
                  <p className="main-input__message">
                    <i className="main-input__message-icon icon icon-info" />
                    85/2000 character maximum
                  </p>
                )}
              >
                <div className="main-input-textarea-wrap mb-0">
                  <TextArea
                    rows={4}
                    placeholder="Describe your listing"
                  />
                </div>
              </Form.Item>
            </Col>
          </Row>
        </Col>
      </Row>
    </Form>
  </LayoutListing>
);

export default ListingDetails;
