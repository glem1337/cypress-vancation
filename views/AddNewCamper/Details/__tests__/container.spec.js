import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import diveTo from 'utils/testHelpers/diveToEnzyme';
import describeValidationSchema from 'utils/testHelpers/describeValidationSchema';
import ROUTES from 'constants/routes';
import { checkPreviousStepCompleteness, fetchCamper } from 'state/concepts/camper/actions';
import { showModal, hideModal } from 'state/modal/actions';
import { createRouteFromPathname } from 'utils/createRouteHelper';
import { loadingSelector } from 'state/data/selectors';

import mockedCamper from '../__mocks__/camper';

import ListingDetails, { ListingDetailsContainer } from '../container';

jest.mock('state/data/selectors', () => ({
  loadingSelector: jest.fn(() => true),
}));

jest.mock('state/concepts/camper/selectors', () => ({
  camperSelector: jest.fn(() => mockedCamper),
  isCamperExistSelector: jest.fn(() => true),
  camperCompletenessSelector: jest.fn(() => 44),
  leavePageMethodSelector: jest.fn(() => jest.fn()),
}));

jest.mock('formik', () => ({
  withFormik: (config) => (Component) => props => {
    const extended = {
      ...config,
      ...props,
      validationSchema: () => config.validationSchema,
    };

    return <Component {...extended} />;
  },
}));

const layoutContainer = (props) => {
  const wrapper = shallow(<ListingDetails {...props} />, { disableLifecycleMethods: true });
  const container = diveTo(wrapper, ListingDetailsContainer);
  const instance = container.instance();

  return {
    wrapper,
    container,
    instance,
  };
};

describe('ListingDetails container tests', () => {
  const store = configureStore()({});
  store.dispatch = jest.fn();
  store.logicMiddleware = {
    whenComplete: jest.fn(),
  };

  const props = {
    store,
    router: {
      push: jest.fn(),
    },
    camperId: 'test id',
    dirty: true,
    isValid: true,
    setFieldValue: jest.fn(),
    handleSubmit: jest.fn(),
    validateForm: jest.fn(),
    values: {
      listingName: '',
      listingDescription: '',
    },
  };

  let wrapper = null;
  let container = null;
  let instance = null;

  beforeEach(() => {
    ({
      wrapper,
      container,
      instance,
    } = layoutContainer(props));

    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('has validation schema', () => {
    expect(describeValidationSchema(wrapper)).toMatchSnapshot();
  });

  it('checks mapPropsToValues configs', () => {
    expect(container.props().mapPropsToValues((container.props()))).toMatchSnapshot();
  });

  it('checks `getInitialProps` static method', async () => {
    const ctx = {
      query: {
        camper_id: 'test id',
      },
      store,
    };

    await ListingDetailsContainer.getInitialProps(ctx);

    expect(ctx.store.dispatch).toHaveBeenCalledWith(
      fetchCamper(ctx.query.camper_id, 'delivery_information'),
    );

    expect(ctx.store.dispatch).toHaveBeenCalledWith(
      checkPreviousStepCompleteness({
        key: ROUTES.ADD_NEW_CAMPER.LISTING_DETAILS.KEY,
        camperId: 'test id',
        ctx,
      }),
    );

    expect(store.logicMiddleware.whenComplete).toHaveBeenCalled();
  });

  describe('tests `onBackButtonClick` instance method', () => {
    it('when form is valid', async () => {
      await instance.onBackButtonClick();

      expect(props.validateForm).toHaveBeenCalled();

      expect(store.dispatch).toHaveBeenCalledWith(
        showModal({
          modalType: 'LEAVE_PAGE_MODAL',
          modalProps: {
            discard: instance.goBack,
            save: instance.saveAndGoBack,
          },
        }),
      );
    });

    it('when form is not valid', async () => {
      container.setProps({
        isValid: false,
      });

      const goBackSpy = jest.spyOn(instance, 'goBack').mockImplementationOnce(() => jest.fn());

      await instance.onBackButtonClick();

      expect(props.validateForm).toHaveBeenCalled();
      expect(store.dispatch).not.toHaveBeenCalled();
      expect(goBackSpy).toHaveBeenCalled();
    });
  });

  it('tests `goBack` instance method', async () => {
    instance.goBack();

    expect(props.router.push).toHaveBeenCalledWith(
      createRouteFromPathname(ROUTES.ADD_NEW_CAMPER.INSURANCE.PATH, props.camperId),
    );

    expect(store.dispatch).toHaveBeenCalledWith(
      hideModal(),
    );
  });

  it('tests `saveAndGoBack` instance method', async () => {
    await instance.saveAndGoBack();

    const route = createRouteFromPathname(
      ROUTES.ADD_NEW_CAMPER.INSURANCE.PATH,
      props.camperId,
    );

    expect(store.dispatch).toHaveBeenCalledWith(hideModal());
    expect(props.setFieldValue).toHaveBeenCalledWith('redirectRoute', route);
    expect(props.handleSubmit).toHaveBeenCalled();
  });

  describe('tests `canSaveAndContinue` instance getter', () => {
    it('when form is not valid', () => {
      const { canSaveAndContinue } = instance;

      expect(canSaveAndContinue).toBe(false);
    });

    it('when form is valid', () => {
      loadingSelector.mockReturnValueOnce(false);

      ({ instance } = layoutContainer({
        ...props,
        values: {
          listingName: 'test',
          listingDescription: '',
        },
      }));

      const { canSaveAndContinue } = instance;

      expect(canSaveAndContinue).toBe(true);
    });
  });
});
