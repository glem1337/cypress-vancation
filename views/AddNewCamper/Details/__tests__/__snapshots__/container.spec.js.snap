// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetails container tests checks mapPropsToValues configs 1`] = `
Object {
  "listingDescription": "desc",
  "listingName": "test",
}
`;

exports[`ListingDetails container tests has validation schema 1`] = `
Object {
  "fields": Object {
    "listingDescription": Object {
      "label": undefined,
      "meta": undefined,
      "notOneOf": Array [],
      "oneOf": Array [],
      "tests": Array [
        Object {
          "name": "max",
          "params": Object {
            "max": 2000,
          },
        },
      ],
      "type": "string",
    },
    "listingName": Object {
      "label": undefined,
      "meta": undefined,
      "notOneOf": Array [],
      "oneOf": Array [],
      "tests": Array [
        Object {
          "name": "max",
          "params": Object {
            "max": 74,
          },
        },
        Object {
          "name": "required",
          "params": undefined,
        },
      ],
      "type": "string",
    },
  },
  "label": undefined,
  "meta": undefined,
  "notOneOf": Array [],
  "oneOf": Array [],
  "tests": Array [],
  "type": "object",
}
`;

exports[`ListingDetails container tests matches snapshot 1`] = `
<ListingDetails
  camper={
    Object {
      "blockedPeriods": Array [],
      "camperPhotos": Array [],
      "camperPolicy": null,
      "deliveryInformation": null,
      "description": "desc",
      "externalCalendars": Array [],
      "id": "a18f7338-c2b3-4cee-be4d-579c0a726c1a",
      "insurance": "pending",
      "name": "test",
      "owner": Object {
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
        "type": "owner",
      },
      "place": null,
      "pricingInfo": null,
      "pricingPeriods": Array [],
      "publicId": "111",
      "specificationDetail": null,
      "status": "draft",
    }
  }
  camperCompleteness={44}
  camperId="test id"
  canSaveAndContinue={false}
  dirty={true}
  handleSubmit={[MockFunction]}
  hideModal={[Function]}
  isCamperExist={true}
  isLoading={true}
  isValid={true}
  mapPropsToValues={[Function]}
  onBackButtonClick={[Function]}
  onSubmit={[Function]}
  router={
    Object {
      "push": [MockFunction],
    }
  }
  setFieldValue={[MockFunction]}
  setLeavePageMethod={[Function]}
  showModal={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "logicMiddleware": Object {
        "whenComplete": [MockFunction],
      },
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  validateForm={[MockFunction]}
  validationSchema={[Function]}
  values={
    Object {
      "listingDescription": "",
      "listingName": "",
    }
  }
/>
`;
